nixos/tests/{k3s,rke2}: merge & cleanup

maevii opened the pull request:
Follow-up on #447847, brings the RKE2 testing infra up to par with K3s and in the process cleans up the K3s tests a bit.

Same reasons as for the previous PR, now that the modules are nearly identical it's possible to test both using a single codebase. The tests themselves should be *mostly* unchanged from K3s pre-PR, though a lot was refactored.

I'd recommend reviewing commit by commit since that shows the actual changes better, the third commit also looks like it has a huge diff because of indentation changes caused by removing `import ../make-test-python.nix (` - nothing except `rancher/default.nix`, `all-tests.nix` and the packages were actually changed.

Sidenote: Since K3s/RKE2 are so similar I'd also like to help with K3s maintenance going forward, do I only need to send a request to join the github team or is there more to it?

Maintainer cc:
@rorosen @zimbatm @stefan-bordei
@NixOS/k3s

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

maevii committed:
nixos/tests/k3s: generalize to rancher

Similarly to 2ce16ee67eae, prepares for merge with nixos/tests/rke2.

maevii committed:
nixos/tests/rke2: merge into nixos/tests/k3s

Follow-up on ac4abf7f7ba9.

azey7f commented:
Sorry for the pings, I think the bot requested everyone for review since this removes an obsolete K3s path from a formatting exception in `ci/default.nix`. CI itself isn't actually touched in any way.

rorosen commented on nixos/tests/rancher/default.nix:
```diff
@@ -0,0 +1,101 @@
+  };
+
+  importTest =
+    path: extraArgs: lib.mapAttrs (_: pkg: import path ((mkTestArgs pkg) // extraArgs)) allPackages;
```
Why even pass `extraArgs`, if it's not used anywhere? What would it be used for?

maevii committed:
nixos/tests/{k3s,rke2}: migrate to runTest

Also exposed a .all attr alongside existing tests, for easily running
all tests for all package versions.

rorosen reviewed (commented):
I love the idea of merging the tests!

I left some minor comments, but overall this looks very good. I think the last push messed up my comments, so some of them may be out of date.

> I'd also like to help with K3s maintenance going forward

Happy to hear that. The header in `team-list.nix` says you should ping one of the team maintainers, if you want to be added.

```
nix eval -f lib teams.k3s.githubMaintainers --json | jq
```

I suggest you open a PR that adds you to the k3s team in `maintainers/github-teams.json` and ping the team maintainers in the PR.

You may also consider adding yourself as a RKE2 maintainer, if you like to.

azey7f commented on nixos/tests/rancher/multi-node.nix:
```diff
@@ -0,0 +1,253 @@
+
+        # wait for the agent to show up
+        server.wait_until_succeeds("kubectl get node agent")
+        server.succeed("kubectl get node >&2")
```
Yep, thanks

azey7f commented:
> Regarding the change of test exposure, what is your use-case to run all tests for all versions?

@rorosen Changed the code to instead expose a .all attr next to the existing tests, should be a good compromise.

wrmilling commented:
@azey7f You should be a member of the team now, I will try and take a look through this PR after work / later this evening.

Mic92 commented:
Can merge when someone verified the tests.

rorosen reviewed (approved):
Thanks!

Changes look good and all test pass for me. Since the change is pretty big, another review would be nice
